agent/agent.go

```diff
@@ -943,6 +951,22 @@ func (a *agent) Close() error {
 	}
 	close(a.closed)
 	a.closeCancel()
+
+	rawMetadata := a.metadata.Load()
+	if rawMetadata == nil {
+		return xerrors.Errorf("no metadata was provided")
+	}
+	metadata, valid := rawMetadata.(codersdk.WorkspaceAgentMetadata)
+	if !valid {
+		return xerrors.Errorf("metadata is the wrong type: %T", metadata)
+	}
+
+	ctx := context.Background()
+	err := a.runShutdownScript(ctx, metadata.ShutdownScript)
+	if err != nil {
+		a.logger.Error(ctx, "shutdown script failed", slog.Error(err))
+	}
```

Comment:
> And also, we might not want to continue on error. Critical workspace tasks could be performed at this step, like syncing filesystem to cold storage or something similar.

I assumed the opposite condition. I wouldn't like to end up with 100 workspaces I can't kill. I suppose that it depends on which option is less evil. I'm fine to change the approach to be aligned with your suggestion.

Reply:
It’s a valid thought. Another way to look at it is that if it’s allowed to fail. Why have it at all? Both are of course valid. In the PR review comment I mentioned a force stop would allow shutting down such workspaces. That’s one option, alternatively it could be a provider option but we should be careful with adding those until there are real use-cases.

Reply:
> In the PR review comment I mentioned a force stop would allow shutting down such workspaces

I might have missed this comment, but do you mean to have the option to configure the forced-stop somewhere in metadata which can be changed using UI or a different place?

Reply:
I was hinting that `coder stop` would receive a new flag, e.g. `--force`. Similarly the WebUI might receive a new button. This would be cli/server-only and not wait for the a-OK from the agent, instead it would perform the terraform teardown immediately.